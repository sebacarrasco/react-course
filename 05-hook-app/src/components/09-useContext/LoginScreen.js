import React, { useContext } from 'react'
import { UserContext } from './UserContext'

export const LoginScreen = () => {

    const { setUser } = useContext(UserContext);
    const user = {
        id: 123,
        name: "Seba"
    }

    return (
        <div>
            <h1> Login Screen </h1>
            <hr />

            <button
                className="btn btn-primary"
                onClick={ () => setUser(user) }
            >
                Login
            </button>
        </div>
    )
}
