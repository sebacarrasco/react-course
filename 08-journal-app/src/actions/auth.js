import Swal from 'sweetalert2';
import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { FinishLoading, StartLoading } from "./ui";
import { notesLogout } from './notes';

export const startLoginEmailPassword = (email, password) => {
    // Cuando una acción retorna un callback (una función)
    // el midleware la ejecutará. Además pasa como parámetro
    // el dispatch gracias a thunk
    return (dispatch) => {

        dispatch(StartLoading());

        // El return es solo para que en testing aparezcan las acciones login y finishLoading como
        // despachadas
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(FinishLoading());
            })
            .catch(error => {
                dispatch(FinishLoading());
                Swal.fire('Error', error.message, 'error');
            });
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        
        dispatch(StartLoading());

        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({ user }) => {
                await user.updateProfile({displayName: name}); // Hay que establecerle el displayName a diferencia
                                                            // de cuando se hace la autenticación con google (se actualiza en db y el objeto)
                dispatch(login(user.uid, user.displayName));
                dispatch(FinishLoading());
            })
            .catch(error => {
                dispatch(FinishLoading());
                Swal.fire('Error', error.message, 'error');
            });
    }
}

export const startGoogleLogin = () => {
    // Como es una tarea asíncrona, se tiene que retornar una función.
    // Nótese que estas acciones son despachadas también.
    // Al ser despachadas y no retornar un objeto sino que una función,
    // entonces se ejecuta la función retornada y se pueden hacer dispatch dentro
    // de la misma función
    return (dispatch) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( userCred => {
                dispatch(login(userCred.user.uid, userCred.user.displayName))
            })
    }
    // El displayName lo tendremos siempre que se autentica con una red social, github o algo así
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});


export const startlogout = () => {
    return async(dispatch) => {
        await firebase.auth().signOut(); //Solo puede fallar por conexión a internet
        dispatch(logout());
        dispatch(notesLogout());
    }
}

export const logout = () => ({
    type: types.logout
});