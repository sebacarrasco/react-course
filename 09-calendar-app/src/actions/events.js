import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";

export const eventStartAddNew = (event) => {
    return async(dispatch, getState) => {

        try {
            
            const response = await fetchWithToken('events', event, 'POST');
            const body = await response.json();

            const { uid, name } = getState().auth;

            if(body.ok)
            {
                event.id = body.event.id;
                event.user = {
                    _id: uid,
                    name
                }
                dispatch(eventAddNew(event));
            }

        } catch (error) {
            console.log(error);
        }

    }
}

export const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActiveEvent = () => ({
    type: types.eventClearActiveEvent
});

export const eventStartUpdate = (event) => {

    return async (dispatch) => {

        try {
            const response = await fetchWithToken(`events/${ event.id }`, event, 'PUT');
            const body = await response.json();
            if (body.ok)
            {
                dispatch(eventUpdated(event));
            }
            else
            {
                Swal.fire("Error", body.msg, "error");
            }
        } catch (error) {
            console.log(error);
        }
    }

}

export const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
});

export const eventStartDelete = () => {

    return async (dispatch, getState) => {

        const { id } = getState().calendar.activeEvent;

        try {
            const response = await fetchWithToken(`events/${ id }`, {},  'DELETE');
            const body = await response.json();
            if (body.ok)
            {
                dispatch(eventDeleted());
            }
            else
            {
                Swal.fire("Error", body.msg, "error");
            }
        } catch (error) {
            console.log(error);
        }
    }

}

export const eventDeleted = () => ({
    type: types.eventDeleted
});

export const eventStartLoading = () => {
    return async(dispatch) => {

        try {
            const response = await fetchWithToken('events');
            const body = await response.json();
            // Pasamos las fechas de string a date
            const events = prepareEvents(body.events);
            dispatch(eventLoaded(events));
        } catch (error) {
            console.log(error);
        }

    }
}

export const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
});

export const eventLogout = () => ({
    type: types.eventLogout
});