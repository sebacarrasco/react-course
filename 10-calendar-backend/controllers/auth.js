const { response } = require('express'); // Solo para la autocmpletacion, no es extrictamente necesario
const User = require("../models/User");
const bcrypt = require("bcryptjs");
const { generateJWT } = require('../helpers/jwt');

// El response = express.response es solo para la autocmpletacion, no es extrictamente necesario
const createUser = async(req, res = response) => {
    // request, response 

    const { email, password } = req.body;

    try {

        let user = await User.findOne({ email });

        // Si user no es null
        if (user)
        {
            return res.status(400).json({
                ok: false,
                msg: "Ese correo ya está asociado a un usuario"
            });
        }

        user = new User(req.body);

        // Encriptar contraseña
        const salt = bcrypt.genSaltSync();
        user.password = bcrypt.hashSync(password, salt);

        await user.save();

        // Generar JWT
        const token = await generateJWT(user.id, user.name);

        // 201 por creado correctamente
        res.status(201).json({
            ok: true,
            msg: 'register',
            uid: user.id,
            name: user.name,
            token
        });
        
    } catch (error) {
        console.log(error);
        // 500 por internal server error
        res.status(500).json({
            ok: false,
            msg: "Por favor hable con el administrador"
        });
    }
}


const loginUser = async(req, res = response) => {
    // request, response 

    const { email, password } = req.body;

    try {
        const user = await User.findOne({ email });

        // Si user es null
        if (!user)
        {
            return res.status(400).json({
                ok: false,
                msg: "Ese correo no está asociado a ningún usuario"
            });
        }

        // Ver si las passwords calzan
        const passwordMatch = bcrypt.compareSync( password, user.password );
        if (!passwordMatch)
        {
            return res.status(400).json({
                ok: false,
                msg: "Contraseña incorrecta"
            });
        }

        // Generar JWT
        const token = await generateJWT(user.id, user.name);
        res.json({
            ok: true,
            uid: user.id,
            name: user.name,
            token
        });

    } catch (error) {
        console.log(error);
        // 500 por internal server error
        res.status(500).json({
            ok: false,
            msg: "Por favor hable con el administrador"
        });
    }
}


const renewToken = async(req, res = response) => {
    // request, response 

    const { uid, name } = req;

    // Generar un nuevo JWT
    const token = await generateJWT(uid, name);
    res.json({
        ok: true,
        token,
        uid,
        name
    });
}


// Recordar que las exportaciones tienen que ser así en node :c
module.exports = {
    createUser,
    loginUser,
    renewToken
}