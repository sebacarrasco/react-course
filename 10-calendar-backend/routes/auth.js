/**
Rutas de Usuarios / Auth
host + /api/auth 
*/

const express = require("express");
const { check } = require("express-validator");
const router = express.Router();

const { createUser, loginUser, renewToken } = require("../controllers/auth");
const { fieldValidator } = require("../middlewares/field-validator");
const { jwtValidator } = require("../middlewares/jwt-validator");

// El primer argumento se concatenará al path que se puso en index,js
// para crear el path real. En este caso es api/auth/new
router.post("/new",
    [ // Arreglo de middlewares que se ejecutarán antes que el controlador
        check("name", "name can't be empty").not().isEmpty(),
        check("email", "email is required").isEmail(),
        check("password", "password should be at least 6 characters long").isLength({ min: 6 }),
        fieldValidator
    ],
    createUser
);

router.post("/",
    [
        check("email", "email is required").isEmail(),
        check("password", "password should be at least 6 characters long").isLength({ min: 6 }),
        fieldValidator
    ],
    loginUser
);

router.get("/renew", jwtValidator, renewToken);

// En node las exportaciones no son iguales :c
module.exports = router;