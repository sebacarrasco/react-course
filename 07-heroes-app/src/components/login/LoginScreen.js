import React, { useContext } from 'react'
import { AuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';

export const LoginScreen = ({ history }) => {

    const { dispatch } = useContext(AuthContext);
    
    const handleLogin = () => {
        // Con el push se añade al historial la siguiente ruta y
        // se puede volver a la página anterior con el botón de atrás (<-).
        // Con el replace se reemplaza, no podrá volver a la página
        // anterior con el botón de atrás (<-).
        // history.push("/");
        // history.replace("/");
        const action = {
            type: types.login,
            payload: {
                name: "Sebastián"
            }
        };
        dispatch(action);

        // Esto es para ir al último path que el usuario visitó
        const lastPath = localStorage.getItem("lastPath") || "/";
        history.replace( lastPath );
    };
    
    return (
        <div className="container mt-5">
            <h1>Login</h1>
            <hr/>

            <button
            className="btn btn-primary"
            onClick={ handleLogin }
            >
                Login
            </button>
        </div>
    )
};