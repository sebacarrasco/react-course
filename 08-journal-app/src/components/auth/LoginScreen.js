import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import validator from 'validator'
import { removeLoginError, setLoginError } from '../../actions/ui';

export const LoginScreen = () => {

    // Este hook es para hacer dispatch de acciones
    // (ya sabe la configuración del store)
    const dispatch = useDispatch();

    // Este hook es para acceder a los states que están en la store
    // Importante mencionar que state es un objeto con tooodos los estados de la store
    const { msgErrorLogin, loading } = useSelector(state => state.ui);

    const [ values, handleInputChange ] = useForm({
        email:'seba@uc.cl',
        password: "123456"
    });

    const { email, password } = values;

    const handleLogin = (e) => {
        e.preventDefault();
        if (isFormValid())
        {
            dispatch(startLoginEmailPassword(email, password));
        }
    }

    const handleGoogleLogin = () => {
        dispatch( startGoogleLogin() );
    }

    const isFormValid = () =>{
        if (!validator.isEmail(email))
        {
            dispatch(setLoginError("Email is not valid"));
            return false;
        }
        else if (password.length <= 5)
        {
            dispatch(setLoginError("Password should be at least 6 characters long"));
            return false;
        }
        dispatch(removeLoginError());
        return true;
    }

    return (
        <>
            <h3 className="auth__title">Login</h3>

            <form
                onSubmit={ handleLogin }
                className="animate__animated animate__fadeIn animate__faster"
            >

                {   
                    // Solo si el error no es null se muestra el div
                    msgErrorLogin && 
                    (<div className="auth__alert-error">
                        { msgErrorLogin }
                    </div>)
                } 

                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    onChange={ handleInputChange }
                    value={ email }
                />
                
                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    onChange={ handleInputChange }
                    value={ password }
                />

                <button
                    type="submit"
                    className="btn btn-primary btn-block"
                    disabled={ loading }
                >
                    Login
                </button>

                <hr />
                
                <div className="auth__social-networks">
                    <p>Login with social networks</p>

                    <div
                    className="google-btn"
                    onClick={ handleGoogleLogin }
                    >
                        <div className="google-icon-wrapper">
                            <img
                                className="google-icon"
                                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                                alt="google button"
                            />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link to="/auth/register" className="link">
                    Create new account
                </Link>

            </form>
        </>
    )
}
