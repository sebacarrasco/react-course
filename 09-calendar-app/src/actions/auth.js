import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch"
import { types } from "../types/types";
import Swal from 'sweetalert2'
import { eventLogout } from "./events";


export const startLogin = (email, password) => {
    
    return async(dispatch) => {

        const response = await fetchWithoutToken('auth', { email, password }, 'POST');
        const body = await response.json();

        if (body.ok)
        {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        }
        else
        {
            Swal.fire('Error', body.msg, 'error');
        }
    }

}

export const startRegister = (email, password, name) => {

    return async(dispatch) => {

        const response = await fetchWithoutToken('auth/new', { email, password, name }, 'POST');
        const body = await response.json();

        if (body.ok)
        {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        }
        else
        {
            Swal.fire('Error', body.msg, 'error');
        }

    }

}

export const startChecking = () => {

    return async(dispatch) => {

        const response = await fetchWithToken('auth/renew');
        const body = await response.json();

        if (body.ok)
        {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        }
        else
        {
            dispatch(checkingFinish());
        }

    }

}

export const checkingFinish = () => ({
    type: types.authCheckingFinish
});

export const login = (user) => ({
    type: types.authLogin,
    payload: user
});

export const startLogout = () => {

    return (dispatch) => {
        localStorage.clear();
        dispatch(eventLogout());
        dispatch(logout());
    }

}

export const logout = () => ({
    type: types.authLogout
})