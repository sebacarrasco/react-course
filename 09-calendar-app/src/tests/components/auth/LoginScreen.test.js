import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../actions/auth';
import Swal from 'sweetalert2';


jest.mock('../../../actions/auth', () => ({
    startLogin: jest.fn(),
    startRegister: jest.fn()
}));

jest.mock('sweetalert2', () => ({
    fire: jest.fn()
}));

// Config de la store
const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <LoginScreen />
    </Provider>
);


describe('Tests for <LoginScreen />', () => {

    beforeEach( () => {
        jest.clearAllMocks();
    });
   
    test('should match the snapshot', () => {
        
        expect(wrapper).toMatchSnapshot();

    });

    test('should dispatch the startLogin action', () => {
       
        wrapper.find("input[name='loginEmail']").simulate('change', {
            target: {
                name: 'loginEmail',
                value: 'test@testing.com'
            }
        });

        wrapper.find("input[name='loginPassword']").simulate('change', {
            target: {
                name: 'loginPassword',
                value: '123456'
            }
        });

        wrapper.find('form').at(0).prop('onSubmit')({
            preventDefault: () => {}
        });

        expect(startLogin).toHaveBeenCalledWith('test@testing.com', '123456');

    });

    test('should not dispatch the startRegister action if passwords do not match', () => {
        
        wrapper.find("input[name='registerPassword1']").simulate('change', {
            target: {
                name: 'registerPassword1',
                value: '123456'
            }
        });

        wrapper.find("input[name='registerPassword2']").simulate('change', {
            target: {
                name: 'registerPasswor2',
                value: '999999'
            }
        });

        wrapper.find('form').at(1).prop('onSubmit')({
            preventDefault: () => {}
        });

        expect(startRegister).not.toHaveBeenCalled();
        expect(Swal.fire).toHaveBeenCalledWith('Error', 'Las contraseÃ±as deben ser iguales', 'error');

    });
    
    test('should dispatch the startRegister action if passwords match', () => {
       
        wrapper.find("input[name='registerPassword1']").simulate('change', {
            target: {
                name: 'registerPassword1',
                value: '123456'
            }
        });

        wrapper.find("input[name='registerPassword2']").simulate('change', {
            target: {
                name: 'registerPassword2',
                value: '123456'
            }
        });

        wrapper.find('form').at(1).prop('onSubmit')({
            preventDefault: () => {}
        });

        expect(startRegister).toHaveBeenCalledWith('', '123456', '');
        expect(Swal.fire).not.toHaveBeenCalled();

    });
    
    

});
