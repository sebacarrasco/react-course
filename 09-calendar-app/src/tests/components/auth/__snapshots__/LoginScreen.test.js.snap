// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <LoginScreen /> should match the snapshot 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Login
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginEmail"
            onChange={[Function]}
            placeholder="Email"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Register
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerName"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerEmail"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword1"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword2"
            onChange={[Function]}
            placeholder="Confirm password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Register"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
